driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -1039,6 +1159,60 @@ public String toString() {
         "Connection[%s, inFlight=%d, closed=%b]", name, inFlight.get(), isClosed());
   }
 
+  static class PortAllocator {
```

Comment:
Isn't it a problem if `startPort` eventually gets high (near `highPort`) and `getNextAvailablePort` won't be able to find a free port? In that case it should probably wrap and start again from `lowPort`.

Also, isn't `busyPorts` array a bit of an overkill? The solution I had in mind was to have a single "global" `lastPort` and `getNextAvailablePort` would bump it higher and higher. By only increasing it, the risk of collisions would be smaller.

I had something like this in mind:

```java
// static for the sake of example
private static final AtomicInteger lastPort = new AtomicInteger(-1);

public static int getNextAvailablePort(int shardCount, int shardId, int lowPort, int highPort) {
    int lastPortValue, foundPort = -1;
    do {
        lastPortValue = lastPort.get();

        // We will scan from lastPortValue
        // (or lowPort is there was no lastPort)
        int scanStart = lastPortValue == -1 ? lowPort : lastPortValue;

        // Round it up to "% shardCount == shardId"
        scanStart += (shardCount - scanStart % shardCount) + shardId;
        assert scanStart % shardCount == shardId;

        // Scan from scanStart upwards to highPort.
        for (int port = scanStart; port <= highPort; port += shardCount) {
            if (isTcpPortAvailable(port)) {
                foundPort = port;
                break;
            }
        }

        // If we started scanning from a high scanStart port
        // there might have been not enough ports left that are
        // smaller than highPort. Scan from the beginning
        // from the lowPort.
        if (foundPort == -1) {
            scanStart = lowPort + (shardCount - lowPort % shardCount) + shardId;
            assert scanStart % shardCount == shardId;

            for (int port = scanStart; port <= highPort; port += shardCount) {
                if (isTcpPortAvailable(port)) {
                    foundPort = port;
                    break;
                }
            }
        }

        // No luck! All ports taken!
        if (foundPort == -1) {
            return -1;
        }
    } while (!lastPort.compareAndSet(lastPortValue, foundPort));

    assert foundPort % shardCount == shardId
            && foundPort >= lowPort
            && foundPort <= highPort;
    return foundPort;
}
```

Reply:
> Isn't it a problem if startPort eventually gets high (near highPort) and getNextAvailablePort won't be able to find a free port? In that case it should probably wrap and start again from lowPort.

For each Connection, `[lowPort, highPort]` range will be search exactly once.

> Also, isn't busyPorts array a bit of an overkill? The solution I had in mind was to have a single "global" lastPort and getNextAvailablePort would bump it higher and higher. By only increasing it, the risk of collisions would be smaller.

Your solution is definitely better in a good case (low % of busy ports), but in a bad case (high % of busy ports), won't it be worse to try to open socket on each port rather than just search an array? Anyway, good case is probably more important, I'll apply your solution (modified a bit, as it needs to take into consideration that there may be multiple sessions).